_extras/refactor-2-code-refactoring.md

```diff
@@ -0,0 +1,411 @@
+
+Firstly, 
+we will modify the function to return the data instead of visualising it because graphs are harder 
+to test automatically (i.e. they need to be viewed and inspected manually in order to determine 
+their correctness). 
+Next, we will make the assert statements verify what the outcome is 
+currently, rather than checking whether that is correct or not. 
+Such tests are meant to 
+verify that the behaviour does not *change* rather than checking the current behaviour is correct
+(there should be another set of tests checking the correctness). 
+This kind of testing is called **regression testing** as we are testing for
+regressions in existing behaviour.
+
+Refactoring code is not meant to change its behaviour, but sometimes to make it possible to verify
+you not changing the important behaviour you have to make small tweaks to the code to write
+the tests at all.
+
+> ## Exercise: Write Regression Tests
+> Modify the `analyse_data()` function not to plot a graph and return the data instead.
+> Then, add a new test file called `test_compute_data.py` in the `tests` folder and 
+> add a regression test to verify the current output of `analyse_data()`. We will use this test
+> in the remainder of this section to verify the output `analyse_data()` is unchanged each time
+> we refactor or change code in the future. 
+> 
+> Start from the skeleton test code below: 
+> 
+> ```python
+> def test_analyse_data():
+>     from catchment.compute_data import analyse_data
+>     path = Path.cwd() / "data"
+>     result = analyse_data(path)
+>
+>     # TODO: add an assert for the value of result
+> ```
+> Use `assert_array_almost_equal` from the `numpy.testing` library to
+> compare arrays of floating point numbers.
+>
+> Remember to run the test using `python -m pytest` from the project base directory:
+> ```bash
+> python -m pytest tests/test_compute_data.py
+> ```
+>
+>> ## Hint
+>> When determining the correct return data result to use in tests, it may be helpful to assert the 
+>> result equals some random made-up data, observe the test fail initially and then 
+>> copy and paste the correct result into the test.
+>>
+>> Remember also that NaN values can be defined using the numpy library (`numpy.nan`).
+> {: .solution}
+>
+>> ## Solution
+>> One approach we can take is to:
+>>  * comment out the visualize method on `analyse_data()` 
+>> (as this will cause our test to hang waiting for the result data)
+>>  * return the data instead, so we can write asserts on the data
+>>  * See what the calculated value is, and assert that it is the same as the expected value
+>> 
+>> Putting this together, your test may look like:
+>>
+>> ```python
+>> import numpy as np
+>> import numpy.testing as npt
+>> from pathlib import Path
+>>
+>> def test_analyse_data():
+>>     from catchment.compute_data import analyse_data
+>>     path = Path.cwd() / "data"
+>>     result = analyse_data(path)
+>>    expected_output = [ [0.        , 0.18801829],
```

Comment:
Putting a note hear to double check this output is correct before merging

Reply:
double check done - this are the correct bunch of random numbers ;-)